deepchecks/utils/gower_distance.py

```diff
@@ -0,0 +1,138 @@
+# ----------------------------------------------------------------------------
+# Copyright (C) 2021-2022 Deepchecks (https://www.deepchecks.com)
+#
+# This file is part of Deepchecks.
+# Deepchecks is distributed under the terms of the GNU Affero General
+# Public License (version 3 or later).
+# You should have received a copy of the GNU Affero General Public License
+# along with Deepchecks.  If not, see <http://www.gnu.org/licenses/>.
+# ----------------------------------------------------------------------------
+#
+"""Module for calculating distance matrix via Gower method."""
+
+import numpy as np
+import pandas as pd
+
+
+def gower_matrix(data: np.ndarray, cat_features: np.array) -> np.ndarray:
+    """
+    Calculate distance matrix for a dataset using Gower's method.
+
+    Gowers distance is a measurement for distance between two samples. It returns the average of their distances
+    per feature. For numeric features it calculates the absolute distance divide by the range of the feature. For
+    categorical features it is an indicator whether the values are the same.
+    See https://www.jstor.org/stable/2528823 for further details.
+    Can deal with missing values.
+    Parameters
+    ----------
+    data: numpy.ndarray
+        Dataset matrix.
+    cat_features: numpy.array
+        Boolean array of representing which of the columns are categorical features.
+
+    Returns
+    -------
+    numpy.ndarray
+     representing the distance matrix.
+    """
+    if not isinstance(data, np.ndarray):
+        data = np.asarray(data)
+
+    feature_ranges = np.ones(data.shape[1]) * -1
+    feature_ranges[~cat_features] = np.nanmax(data[:, ~cat_features], axis=0) - np.nanmin(data[:, ~cat_features],
+                                                                                          axis=0)
+
+    result = np.zeros((data.shape[0], data.shape[0]))
+    for i in range(data.shape[0]):
+        for j in range(i, data.shape[0]):
+            value = calculate_distance(data[i, :], data[j, :], feature_ranges)
+            result[i, j] = value
+            result[j, i] = value
+
+    return result
+
+
+def gower_matrix_n_closets(data: np.ndarray, cat_features: np.array, num_neighbours: int):
+    """
+    Calculate distance matrix for a dataset using Gower's method.
+
+    Can deal with missing values.
+    See https://statisticaloddsandends.wordpress.com/2021/02/23/what-is-gowers-distance/ for further details.
```

Comment:
If this is the "entry" function for gower this needs to be a bit fuller (and mainly explain the "trick" here of saving only the num_neighbours per sample.